docs/cookbook/knowledge_graph.md

```diff
@@ -0,0 +1,211 @@
+
+## Setup
+
+```python
+pip install mirascope[openai]
+# (Optional) For visualization
+pip install matplotlib, networkx
+```
+
+## Create the `KnowledgeGraph`
+
+The first step is to create a `KnowledgeGraph` with `Nodes` and `Edges` that represent our entities and relationships. For our simple recipe, we will use a Pydantic `BaseModel` to represent our `KnowledgeGraph`:
+
+```python
+from pydantic import BaseModel, Field
+
+
+class Edge(BaseModel):
+    source: str = Field(..., description="The source node of the edge")
+    target: str = Field(..., description="The target node of the edge")
+    relationship: str = Field(
+        ..., description="The relationship between the source and target nodes"
+    )
+
+
+class Node(BaseModel):
+    id: str = Field(..., description="The unique identifier of the node")
+    type: str = Field(..., description="The type or label of the node")
+    properties: dict | None = Field(
+        ..., description="Additional properties and metadata associated with the node"
+    )
+
+
+class KnowledgeGraph(BaseModel):
+    nodes: list[Node] = Field(..., description="List of nodes in the knowledge graph")
+    edges: list[Edge] = Field(..., description="List of edges in the knowledge graph")
+```
+
+Our `Edge` represents connections between nodes, with attributes for the source node, target node, and the relationship between them. While our `Node` defines nodes with an ID, type, and optional properties. Our `KnowledgeGraph` then aggregates these nodes and edges into a comprehensive knowledge graph.
+
+Now that we have our schema defined, it's time to create our Agent.
+
+## Creating the `Agent`
+
+We first define a Pydantic `BaseModel` as our `Agent` with a `knowledge_graph` property. This will be the state of the agent. We first define our tool `_generate_knowledge_graph` which will create a knowledge graph based on our document and user query. We are taking this unstructured [Wikipedia](https://en.wikipedia.org/wiki/Large_language_model) article and creating a knowledge graph.
+
+```python
+from mirascope.core import openai, prompt_template
+
+
+class Agent(BaseModel):
+    knowledge_graph: KnowledgeGraph = KnowledgeGraph(nodes=[], edges=[])
+
+    @openai.call(model="gpt-4o-mini", response_model=KnowledgeGraph)
+    @prompt_template(
+        """
+        SYSTEM:
+        Your job is to create a knowledge graph based on the text and user question.
+        
+        The article:
+        {text}
+
+        Example:
+        John and Jane Doe are siblings. Jane is 25 and 5 years younger than John.
+        Node(id="John Doe", type="Person", properties={{"age": 30}})
+        Node(id="Jane Doe", type="Person", properties={{"age": 25}})
+        Edge(source="John Doe", target="Jane Doe", relationship="Siblings")
+
+        USER:
+        {question}
+        """
+    )
+    def _generate_knowledge_graph(self, question: str) -> openai.OpenAIDynamicConfig:
+        text = ""
+        with open("wikipedia.txt") as f:
```

Comment:
as you mention, this is somewhat silly. Instead, have the argument be the url / local file path and then open it in the call (that way it's actually dynamic)

Reply:
yeah that makes sense